### ISSUE TITLE
Add some sort of "whyareyourunning" functionality

### ISSUE THREAD

**Issue by SUPERCILEX** [gradle/gradle#6211]
Proguard has a "whyareyoukeeping" flag which lets devs figure why something is happening. It would be super helpful to learn about why a task is being run, aka its dependencies. Maybe provide this in `--scan`?

**Comment by big-guy** [comment]
`--info` and `--scan` does this.

- If outputs were removed, we report that the outputs are missing.
- If the inputs have changed, we report that the inputs have changed down to the property level (for non-FileCollection types).
- If the input _files_ have changed, we report a subset of the changes to the input files (removed, changed, added).

This is visible in build scans and when comparing builds.  This build failed:
https://enterprise-samples.gradle.com/s/htmj5xe4uxl3k/timeline?task=3scflzwkg4yxw

And we can see `compileJava` runs because `Input property 'classpath' file repos/repo2/com/acme/utils/1.6/utils-1.6.jar has changed`.  

In the build comparison (with a working build), we can see the successful build has a different classpath for `compileJava`: 
https://enterprise-samples.gradle.com/c/b5hkvdpvrqinu/htmj5xe4uxl3k/task-inputs

`--info` shows a subset of this information right before the task in question is executed.

**Comment by SUPERCILEX** [comment]
No, sorry for not being clearer: I'm asking to be able to see which task dependencies caused it to run. Example: I run assemble, but task x runs. I'd want to see that x ran because assemble depends on it.
